### ISSUE TITLE
WebSockets don't work on Apache

### ISSUE THREAD

**Issue by FooBarWidget** [phusion/passenger#1202]
If the response has the chunked transfer encoding, then WebSockets handshakes become corrupted. See https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/phusion-passenger/XtjxA8E1idw/Z029OysamjsJ for the report.

**Comment by kiranos** [comment]
any news on this?

**Comment by FooBarWidget** [comment]
Unfortunately not. If you need WebSockets, for the time being we recommend using Nginx, or using Passenger Standalone behind Apache + mod_proxy.

**Comment by vanboom** [comment]
See https://github.com/phusion/passenger-ruby-websocket-demo/issues/2, appears WebSockets are still not working under Apache mod_passenger  in 5.0.18?  (It's very possible that I am doing something wrong in my Apache .conf)  Thanks for Passenger - it's incredible!

**Comment by FooBarWidget** [comment]
@vanboom That is correct, WebSockets do not work yet in the Apache version. For the time being, we recommend that you use the Nginx version, or Apache + Standalone in a reverse proxy setup, if you need WebSockets.

**Comment by ThaddeusAid** [comment]
It has been almost a year since the last update. Did the upgrade to apache 2.4.x improve this?

**Comment by CamJN** [comment]
There is a [websocket proxy module](https://httpd.apache.org/docs/2.4/mod/mod_proxy_wstunnel.html) now for Apache 2.4, however we have not found it to be robust enough to recommend yet.

**Comment by mk-pmb** [comment]
Just discovered passenger, sounds like a gift from heaven. :+1:
In the project I want to use to explore passenger, I can set specific ports (or env var names to read the ports from) for WebSockets and EventSource, will that help? It's meant as a work-around to use with Apache's mod_proxy, but maybe passenger can intercept to sanitize and monitor stuff.

**Comment by mk-pmb** [comment]
Just for the record, some people seem to have managed to [use websockets in apache with wstunnel](http://stackoverflow.com/questions/27526281/websockets-and-apache-proxy-how-to-configure-mod-proxy-wstunnel).

**Comment by CamJN** [comment]
Yes, that's the websocket proxy module I mentioned, however it only works for some people. We'll consider recommending it when it is more robust.

**Comment by jamgregory** [comment]
@CamJN - out of interest, what issues did you have with the websocket module in Apache? I presume there's some work involved on the Passenger side to integrate (if possible) with that so ActionCable will work through Passenger directly. We're looking at a future development project that will require ActionCable and ideally we'd like Passenger to support this if possible.

**Comment by CamJN** [comment]
As one example since this module is just a websocket-aware proxy, you need proxy support in your websocket library so it won't just work for everyone.

**Comment by mk-pmb** [comment]
> you need proxy support

What features will that consist of?
Can I test all those features on a per-app level, independent of which ws lib it uses?
Would it helpt to have a stub option where I can configure what work-arounds to use until some feature is available?

One feature I might imagine as "proxy support" would be about determining the IP from which a connection arrives. A work-around might be to specify the network interface from which the proxy should send its packets, to ensure it uses an interface (and thus IP address) which the app doesn't assume as privileged in any way. The setting might have an optional "allow localhost" flag which determines whether to fail loudly in case the specified interface will be undistinguishable from localhost.

**Comment by CamJN** [comment]
> What features will that consist of?

Generating urls for links and js that refer to the proxy's url/port not the app's.

> Can I test all those features on a per-app level, independent of which ws lib it uses?

No, because they have to come from the ws lib.

> Would it help to have a shims option where I can configure what workarounds to use until some feature is available?

Yes, but I don't know if such a thing exists.

**Comment by mk-pmb** [comment]
I'll share my idea since I couldn't finish my research on it: A lot of web apps have their websockets behind very few, short URL prefixes like `/ws/` (https://api.example.com/ws/chat). If we can make the browser and potential proxyies send the websocket request as the first request in the TCP connection, a netfilter string match or `l7-filter` should be able to pick up on the request URL and NAT it to bypass Apache.

**Comment by blevine** [comment]
Anyone having any luck with this?  mod_wstunnel has been around for a while now.  We've had no problem getting websockets to work over an Apache reverse proxy using mod_wstunnel.  However, still not sure how to get this to work using Passenger in Apache integration mode.

**Comment by heangratha** [comment]
Any one have luck to get websocket work with Apache?
This is my Apache configure get from:
https://www.phusionpassenger.com/library/walkthroughs/deploy/meteor/ownserver/apache/oss/xenial/deploy_app.html#edit-apache-configuration-file

But the websocket won't work.. I'm using Ubuntu 16.04, Apache 2.4.18, Phusion Passenger 5.2.0

**Comment by jamgregory** [comment]
@heangratha - the only solution we could come up with was to run a separate instance of the app in Puma, then proxy all WebSocket requests to that using mod_wstunnel.

**Comment by P9GIT** [comment]
Any movement on this issue?

**Comment by CamJN** [comment]
Not right now, it is unfortunately not a priority at the moment. People are recommended to use nginx if they need websockets, as it's still a pain with Apache.

**Comment by P9GIT** [comment]
@CamJN 

OK, I see. Thank you for the update.

**Comment by mk-pmb** [comment]
Update on my research: We could probably write a module like fdpass to get the socket descriptors. We'd then have to exfiltrate the connection's SSL context and send that along with the connection socket file descriptor to a server that can receive both and continue the SSL session. The metasploit codebase probably has useful parts to reuse here.

> Generating urls for links and js that refer to the proxy's url/port not the app's.

I'm optimistic that URL rewriting might be an option for people with a bit of backgorund knowledge about the app – which the app's admin should have.

**Comment by siegy22** [comment]
Is this a wontfix?

**Comment by CamJN** [comment]
@siegy22 not exactly, but it's blocked on the Apache project improving their module.

**Comment by annymosse** [comment]
> @vanboom That is correct, WebSockets do not work yet in the Apache version. For the time being, we recommend that you use the Nginx version, or Apache + Standalone in a reverse proxy setup, if you need WebSockets.

@FooBarWidget 
how can i do it (Apache + Standalone in a reverse proxy setup)?"still newer in servers configs"

**Comment by mk-pmb** [comment]
@annymosse
[How to install Passenger Standalone](https://www.phusionpassenger.com/library/install/standalone/install/oss/)
Once you have that running, it gives a web server. You'll want to ensure it is not reachable directly from the outside. To achieve this, you can use explicit interface binding (to a non-public interface) and/or firewall rules.
The [Apache Reverse Proxy Guide](http://httpd.apache.org/docs/2.4/howto/reverse_proxy.html) shows how to make a publicly reachable apache forward traffic for some parts of your URL space, to the guarded private Passenger server.

**Comment by beppe9000** [comment]
Does this apply to cpanel's apache/phusion ?

**Comment by deltafactory** [comment]
@beppe9000 Yes, since cPanel's default server is Apache.

**Comment by mjobin-mdsol** [comment]
@CamJN 
> not exactly, but it's blocked on the Apache project improving their module.

is there an open issue on Apache's side we can track?

**Comment by mjones-vsat** [comment]
Any updates on this?

**Comment by mk-pmb** [comment]
It should now be easier because afaik modern Apache no longer supports threaded MPM.

**Comment by mk-pmb** [comment]
Sorry for the wrong info earlier. Seems like apache itself still supports prefork worker mpm, just my hosting management software doesn't.

**Comment by mjones-vsat** [comment]
Is there some Apache issue that we can refer to that will unblock this?

**Comment by omar391** [comment]
how about this: https://httpd.apache.org/docs/2.4/mod/mod_proxy_wstunnel.html

**Comment by CamJN** [comment]
mod proxy wstunnel forces you to specify a separate target for your ws traffic. Essentially you need something listening on another port than the main http port, which will handle your ws traffic. I'm fairly certain that limitation is why people want a better solution.

**Comment by strugee** [comment]
I don't believe that's true. You don't need a separate port, just a separate path. This snippet, for example, is working just fine for me in production:

```apache
# Proxy WebSocket requests to /stream
ProxyPass /stream ws://127.0.0.1:56488/stream retry=0 timeout=5

# Proxy all other requests requests to /
ProxyPass / http://127.0.0.1:56488/ retry=0 timeout=5
ProxyPassReverse / http://127.0.0.1:56488/
```

**Comment by mjobin-mdsol** [comment]
what ? I can use Apache with Rails ActionCable now ?

@strugee you are a life saver... I had long forgotten about this...

**Comment by sudoremo** [comment]
> I don't believe that's true. You don't need a separate port, just a separate path. This snippet, for example, is working just fine for me in production:
> 
> ```apache-config
> # Proxy WebSocket requests to /stream
> ProxyPass /stream ws://127.0.0.1:56488/stream retry=0 timeout=5
> 
> # Proxy all other requests requests to /
> ProxyPass / http://127.0.0.1:56488/ retry=0 timeout=5
> ProxyPassReverse / http://127.0.0.1:56488/
> ```

Just to clarify: The actual WebSockets-endpoint is also delivered by Passenger, and not a separate process? In the past, I had to use a separate `puma` process, on a separate port, to facilitate WS with Apache.

**Comment by strugee** [comment]
I can't comment on Passenger support as I haven't used it for years. I was talking strictly about generic Apache mod_proxy and mod_proxy_ws.

**Comment by Gu7z** [comment]
Almost 10 years, has there been any progress or alternative solutions explored over, besides resorting to running a parallel server and rewriting routes in Apache?

**Comment by mjobin-mdsol** [comment]
I'm sure very sad when I have to move away from the Apache module when an app starts using WebSocket.

**Comment by slopez31** [comment]
any update ?
